node/src/components/block_accumulator.rs

```diff
@@ -227,37 +245,65 @@ impl BlockAccumulator {
+        // If the acceptor already exists, just register the peer, if applicable.
+        let entry = match self.block_acceptors.entry(block_hash) {
+            btree_map::Entry::Occupied(entry) => {
+                if let Some(sender) = maybe_sender {
+                    entry.into_mut().register_peer(sender);
+                }
+                return;
+            }
+            btree_map::Entry::Vacant(entry) => entry,
+        };
+
+        // The acceptor doesn't exist. Don't create it if the item's era is not provided, the local
+        // tip doesn't have an era or the item's era is older than the local tip era by more than
+        // `recent_era_interval`.
+        match (maybe_era_id, self.local_tip) {
+            (Some(era_id), Some(local_tip))
+                if era_id >= local_tip.era_id.saturating_sub(self.recent_era_interval) => {}
+            _ => return,
+        }
+
+        // Check that the sender isn't telling us about more blocks than expected.
+        if let Some(sender) = maybe_sender {
+            let block_timestamps = self.peer_block_timestamps.entry(sender).or_default();
+
+            // Prune the timestamps, so the count reflects only the most recently added acceptors.
+            let purge_interval = self.purge_interval;
+            while block_timestamps
+                .front()
+                .map_or(false, |(_, timestamp)| timestamp.elapsed() > purge_interval)
             {
-                let acceptor = self
-                    .block_acceptors
-                    .entry(block_hash)
-                    .or_insert_with(|| BlockAcceptor::new(block_hash, None));
-                acceptor.register_peer(sender);
+                block_timestamps.pop_front();
             }
-            // In all other cases (i.e. the item's era is not provided, the
-            // local tip doesn't have an era or the item's era is older than
-            // the local tip era by more than `recent_era_interval`), we only
-            // register the peer if there is an acceptor for the item already.
-            _ => {
-                if let Some(acceptor) = self.block_acceptors.get_mut(&block_hash) {
-                    acceptor.register_peer(sender)
-                }
+
+            // Assume a block time of at least 1 millisecond, so we don't divide by zero.
+            let block_time = self.block_time.max(TimeDiff::from(1));
+            let expected_blocks = (purge_interval / block_time) as usize;
+            let max_block_count = PEER_RATE_LIMIT_MULTIPLIER.saturating_mul(expected_blocks);
+            if block_timestamps.len() >= max_block_count {
+                warn!(
+                    ?sender, %block_hash,
+                    "rejecting block hash from peer who sent us more than {} within {}",
+                    max_block_count, self.purge_interval,
+                );
+                return;
             }
+            block_timestamps.push_back((block_hash, Timestamp::now()));
```

Comment:
In general, I like to lift determination of the current timestamp up as much as possible, since every single one of these is a syscall and it makes testing a headache. I think it's fine here though, since we never upsert more than one thing at a time?

Reply:
Yes, there are currently tons of `Timestamp::now()` in the new 1.5 code. Maybe needs discussion.